### PR TITLE
[READY] I got guns

### PR DESCRIPTION
**Pull request by actioninja**
:cl: actioninja\memager
refactor: Ballistic guns have been almost entirely reworked from a backend side. This primarily focused on moving as much as possible into the base ballistic gun and turning as many other guns into essentially varedits as possible.
add: Guns can now be racked. This is the default action unless the magazine is empty.
add: Guns now can have the magazine removed by clicking on them with an open hand while they are in hand.
tweak: Suppressors are now removed by alt clicking instead of clicking with an open hand
add: Guns now have various bolt types that all function a bit differently. Open bolts cannot have a bullet chambered with no mag, locking bolts lock back after running out of ammo, etc.
add: All ballistic gun sounds are controlled by variables instead of hardcoded usages.
tweak: The l6 LMG has a reworked control scheme. alt + click to open and close the cover, open hand to remove mag, place mag in by slapping them together, default action is rack.
tweak: Functionality that used to be snowflake code such as tactical reloads has been moved to the base gun, and can be toggled by variables.
refactor: All shotguns are now properly subpaths of the shotgun type. They still work the same.
imagedel: Gun sprites have been almost entirely overhauled to use overlays instead of states. This collapsed the L6 sprite from 20 sprites to 9 sprites to give an example.
soundadd: Remixed versions of the shotgun and base gun firesound
soundadd: New rifle firesound for l6 and moist nugget, new SMG fire sound, new sniper fire sound.
soundadd: Lots of new weapon operation sounds such as racking and bolts and the like.
imageadd: New sprites for the Riot Shotgun, Combat Shotgun, c20r, Deagle, m90, double barrel shotgun and 1911 by Memager.
imageadd: All gun sprites that were pointing to the left have been flipped to point to the right.
balance: Shotguns now can be pumped faster.
balance: Bulldog can now be tac reloaded
balance: Sawn off shotguns now have an accuracy penalty and recoil
/:cl:

[GUNS](https://www.youtube.com/watch?v=og243Dom4Sw)

This is probably way too much in one PR but it's all fairly related so fuck it. Complain if it's too much and I'll split everything that's not the functionality changes off.

Sprites:
image

Should I record a video for all the sounds? It's like 10 minutes of effort man...

Shit I still need to do:
- [x] FIX THE DOUBLE BARREL (AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA)
- [x] Fix the bulldog's magazine sprites in a way that doesn't result in snowflake shitcode
- [x] Fix the moist nugget
- [x] Make tac reloads not dump the mag on the floor.
- [x] Figure out what the hell to do with cursed bolt action and arcane barrage
- [x] Decide what to do about the L6's control scheme.
- [x] Make locking bolts actually fucking lock, ebic shitcode
- [x] Make revolvers able to be spun at any time
- [x] Convert every sprite to new formats
- [x] Make sawn weapons shittier in some way.
- [x] More sounds
- [x] More sprites
- [x] Fix the various typos I have introduced.
- [x] Bugtest this shitcode
- [x] ~~reenable the laser gatling~~

### CODE REVIEW

**Comment by davethwave** [comment]
Hope this goes well.

**Comment by zamolxius** [comment]
this is awesome. can you let us screwdriver energy guns to pulse/cut various wires for various effects? i.e focus the beam for a short ranged melee weapon, unfocus it for flashlight/flash, make it a shotgun, make it fire on demand with a signaller (store player facing when dropping them, directional item on floor)

**Comment by MACIEKBAKI** [comment]
make it so you can upgrade energy gun's batteries.

**Comment by Steelpoint** [comment]
This is an interesting set of changes, all over I think they will do well. I like the sprite changes but that's a subjective matter.

I'm loath to suggest features for someone else's PR, but one concept I can think of is to add advantages and disadvantages to 'sawing' off a shotgun (Riot and Double Barrel), maybe even make it a concept where you can fold the stock of those shotguns to make them more compact but at some other cost, maybe lower fire rate or wider pellet spread.

**Comment by JStheguy** [comment]
You might want to check over the deagle sprites again, you left the camo and gold deagles untouched and you haven't even changed the deagle magazine overlay to fit with the new sprite.

**Comment by Potato-Masher** [comment]
Geez, the gun refactor PR to end all others.

Cool stuff, though I'll have to think more on the riot shotgun sprite. Not sure if I prefer the old one or the new one.

**Comment by actioninja** [comment]
@zamolxius @MACIEKBAKI This pr is entirely about ballistic guns, I am trying to avoid touching anything else at all.
@Steelpoint don't worry about it, this is intended to add as much functionality in one pass on ballistic guns in one pass as possible. I'm definitely considering things like stat mods for sawn weapons, that's a solid idea. Currently there's no downside for most weapons to chopping them off.
@JStheguy I'm aware of most of the sprite issues like this, I still need to do a
final pass on most of them.
@subject217 To argue with you a bit, not all composite guns have grey bodies. The P90 is actually a good example of a composite gun with a black body.
Whatever the current riot shotgun sprite is, it doesn't look parkerized, it looks like a grey mess. The problem is if you make it just black like most actual shotguns it loses it's easy visual distinction. Memager roughly made the sprite based on the Model 37 from Alien Isolation, and I thought even if the white/chromed barrel doesn't make a whole lot of sense on a practical riot weapon, it at least is visually distinct from the combat shotgun.

**Comment by subject217** [comment]
It's not about the color, it's about the shading. Look at the whole hand grip section of the "new" C-20r. What is that shading? It doesn't make any sense.

**Comment by actioninja** [comment]
the fucking mapdiff render broke and made the checks fail REEE

**Comment by Eskjjlj** [comment]
Now you won't be able to reload your gun if you only have one hand. Huge nerf.

**Comment by actioninja** [comment]
You still can. The default action is rack _unless_ the magazine is empty.

**Comment by actioninja** [comment]
If I do change LMG reloads, I definitely wouldn't be making it easier.
The whole "shotgun pump speed" change is due to shotgun pumping fundamentally working differently. I have vars in place that could revert to the old speed, but comments in the code indicated the delay was to prevent people from spamming pump and lagging the server. So I gave it the default rack timer instead of the old pump delay, and testing it in game it felt better. It was easier to quickly unload a shotgun and also made them feel a little less clunky to use.
I also figured it was kind of an inconsequential change. I really thought about it, and you're not straight pumping shots out of a shotgun as fast as you can unless you're spamming a hallway, and either way the person on the receiving end is going to get fucked up if you're doing that.
But whatever. I might end up splitting any actual balance changes off from this PR anyways. That's not that horrible of an idea.

**Comment by Supermichael777** [comment]
Make cutdown riot shotgun have a long shot delay (not as spamable). make Double barrel weaken its bullets ~20-30%.

The assault shotgun (short combat shotgun) just needs the lower mag capacity.

**Comment by Poojawa** [comment]
I don't understand the logic behind sawing off riot and combat shotties.

You realize that it would just make them single shot weapons, what with the ammo tube and the spring inside being sawn off in the process.

**Comment by actioninja** [comment]
> Make cutdown riot shotgun have a long shot delay (not as spamable). make Double barrel weaken its bullets ~20-30%.
> 
> The assault shotgun (short combat shotgun) just needs the lower mag capacity.
> 
> Though mixing balance and refactor is bad
I'm beyond all reason at this point.
This isn't just refactor though, there's a decent amount of functionality change as well. This is turning into such a fucking headache that I'm debating completely splitting off anything that's not the core functionality changes because it's making me want to commit die.

> I don't understand the logic behind sawing off riot and combat shotties.
> 
> You realize that it would just make them single shot weapons, what with the ammo tube and the spring inside being sawn off in the process.
The sprites I'm using are just the full length ones without stocks, not true sawn off. Having a reduced ammo capacity doesn't make much sense in that case, but is fair enough.
Also the combat shotgun you can't even saw off, it's just a compact version of that the warden starts with.

**Comment by subject217** [comment]
the old sawn off riot shotgun sprite just removed the stock.

**Comment by SpaceManiac** [comment]
Your PR has been marked Work In Progress for nine days with no updates. It may be marked stale if you do not continue work or otherwise provide a status update.

**Comment by actioninja** [comment]
I've been super busy the past week or so. after tomorrow I should be able to work on this more.

**Comment by actioninja** [comment]
this pr is a fucking mess and I got no clue how to go about fixing it without starting over.
So that's what I'm going to do.

**Comment by subject217** [comment]
i believe in you actioninja, you can finish this

**Comment by davethwave** [comment]
Good luck

**Comment by actioninja** [comment]
revenge of the shit gun pr: I fixed double barrels this time edition

**Comment by actioninja** [comment]
How do I fix the fucked map build

**Comment by Cyberboss** [comment]
Yell at me more

**Comment by subject217** [comment]
get 👏 this 👏 merged 👏

**Comment by actioninja** [comment]
I'm still working on sounds and sprites but this could use a code review now.

**Comment by actioninja** [comment]
hey the map diff is finally fixed

**Comment by optimumtact** [comment]
are you still working on this? also why not in the future make the pr when it's ready, rather than halfbaked

**Comment by VileBeggar** [comment]
i love you

**Comment by subject217** [comment]
oh God, here we go

File: code/modules/projectiles/guns/ballistic/automatic.dm [hunk 1]
```diff
@@ -235,116 +204,77 @@
...
-		magazine.update_icon()
-		magazine.forceMove(drop_location())
-		user.put_in_hands(magazine)
-		magazine = null
-		update_icon()
-		to_chat(user, "<span class='notice'>You remove the magazine from [src].</span>")
-		playsound(user, 'sound/weapons/magout.ogg', 60, 1)
+		return
+	if (!cover_open)
+		to_chat("<span class='warning'>[src]'s cover is closed! Open it before trying to remove the magazine!</span>")
```
Comment (yoyobatty) [inline code/modules/projectiles/guns/ballistic/automatic.dm]: You forgot to include the mob argument, which would be user in this case. This message fails to show completely.

Comment (actioninja) [reply code/modules/projectiles/guns/ballistic/automatic.dm]: FUCK

File: code/modules/projectiles/guns/ballistic.dm [hunk 2]
```diff
@@ -3,77 +3,212 @@
...
+		magazine = null
+	if (display_message)
+		to_chat(user, "<span class='notice'>You pull the [magazine_wording] out of \the [src].</span>")
+	update_icon()
 
 /obj/item/gun/ballistic/can_shoot()
-	if(!magazine || !magazine.ammo_count(FALSE))
-		return 0
-	return 1
+	return (chambered || magazine || magazine.ammo_count(FALSE))
```
Comment (yoyobatty) [inline code/modules/projectiles/guns/ballistic.dm]: should put a ?. check for magazine

**Comment by GravitronBlue** [comment]
image
Missing a sprite change in [icons\mob\back.dmi]
_(Immersion crushing shotgun shells)_

**Comment by subject217** [comment]
make an issue report

**Pull request merged**